### PR TITLE
Find all permutation of a string using recursion

### LINKED ISSUE #3010: All permutation of a string
## 💥 Proposal

(A clear and concise description of what the proposal is.)

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

(Write your answer here.)
Yes, I have read the contributing guidelines.
hey, I am Kirti as a GSSoC '21 participant,
Through this method we can find all permutation of a string. for example if given string is "abc" then output will be ("abc","acb","bac","bca","cab","cba").
Please assign this task to me!

### PR DESCRIPTION
**Pull request by Kirti1807**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

### Description

(Write your answer here.)

### Checklist

- [ Yes] I've read the contribution guidelines.
- [ Yes] I've checked the issue list before deciding what to submit.
- [ yes] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)
In this I try to find all permutation of a string using recursion.
Fixes #3010

### CODE REVIEW

**Comment by Kirti1807** [comment]
@ankitaggarwal23  sir I resolve merge conflict issue and all other issue whichever I noted in my PR. Can you review my PR.

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/*
+    Here I use recursion to find all permutation of a string.
+    This method is useful for small length's strings because its time complaxity is in terms of n!  
+    and for non repeated character in string.
+*/
+#include<bits/stdc++.h>
+using namespace std;
+
+void rec(string s , int left_index , int right_index){
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Use proper name to define function. (Function name is not understandable. What is `rec`).

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    // here l is starting index and r is last index of string
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Remove redundant comments.

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    if( left_index == right_index ) {
+        // this is base case where recursion is going to stop
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Remove redundant comments.

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        cout<<s<<"  ";
+        
+    }
+    // here my approach is to fix a character and find all permutation of next all character 
+    // and fixing of characters is done by swaping
+    for(int i = left_index ; i <= right_index ; i++) {
+        swap(s[left_index] , s[i]) ;
+        rec(s , left_index + 1 , right_index) ; // after fixing of characters we find all permutation of rest all characters by calling same function
+        swap(s[left_index] , s[i]) ;  // here we get again initial string and again this process is start
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Remove redundant comments.

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+}
+
+int main() {
+    string S ;
+    cin>>S;
+    rec(S , 0 , S.size() - 1 ) ;
+
+}
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Remove extra line.

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+/*
+    Here I use recursion to find all permutation of a string.
+    This method is useful for small length's strings because its time complaxity is in terms of n!  
+    and for non repeated character in string.
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: - Use multiline comments.
- Add a new line above `bits/c++.h`

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+int main() {
+    string S ;
+    cin>>S;
+    rec(S , 0 , S.size() - 1 ) ;
+
+}
+
+/*
+    Input_1:xyz
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Above your Input_1 just add `SAMPLE OUTPUTS:`

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by Kirti1807** [comment]
@ankitaggarwal23 
Sir , I make all changes, can you review again .

File: C-Plus-Plus/cp/recursion_string_permutation.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/* Here I use recursion to find all permutation of a string.
+This method is useful for for non repeated character in string and small length's strings because its time complaxity is in terms of n! .  */
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/cp/recursion_string_permutation.cpp]: Use multiline comments for description `//`. Rest looks good to merge.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Review by ankitaggarwal23 — APPROVED** [review]

File: C-Plus-Plus/cp/recursion_string_permutation.cpp (hunk 8 above)
Comment (Kirti1807) [reply C-Plus-Plus/cp/recursion_string_permutation.cpp]: @ankitaggarwal23 
Sir , changes done please can you review again .

**Comment by Kirti1807** [comment]
Thank you Sir !

**Pull request merged**